web/models/key.ex

```diff
@@ -31,11 +33,35 @@ defmodule HexWeb.Key do
   end
 
   def all(user) do
-    assoc(user, :keys)
+    from(k in assoc(user, :keys), where: is_nil(k.revoked_at))
   end
 
   def get(name, user) do
-    from(k in assoc(user, :keys), where: k.name == ^name)
+    from(k in assoc(user, :keys), where: k.name == ^name and is_nil(k.revoked_at))
+  end
+
+  def get_revoked(name, user) do
+    from(k in assoc(user, :keys), where: k.revoked_name == ^name)
+  end
+
+  def revoke(user, key_name, revoked_at \\ Ecto.DateTime.utc) do
+    from(k in assoc(user, :keys),
+      where: k.name == ^key_name and is_nil(k.revoked_at),
+      update: [set: [
+        revoked_name: k.name,
+        revoked_at: fragment("?", ^revoked_at),
+        name: nil
+      ]])
+  end
```

Comment:
I originally had it using a changeset, but realized there could be a case where two delete requests come in near the same time and one succeeds while the other causes a crash on the server-side.

My thought was this would be the closest thing to an atomic operation we could get.  Unless I'm missing something and changesets are also able to offer a similar guarantee.